server/license: Flow cluster init timestamp through tenant connector

spilchen opened the pull request:
Currently, secondary tenants cannot access the cluster initialization grace period timestamp in the KV store because they lack permission to access the system keyspace. Instead of reading it directly from the KV, we now pass the timestamp through the tenant connector, making it accessible to secondary tenants. The timestamp is only written once during the system tenant’s initial startup, so it remains a constant value from the perspective of secondary tenants. The value is sent from the system tenant to secondary tenants using the existing TenantSettingsRequest for METADATA_EVENT.

This change will be backported to versions 24.2, 24.1, 23.2, and 23.1.

Epic: CRDB-39988
Closes CRDB-42309
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131750)
<!-- Reviewable:end -->

rafiss reviewed (approved):
nice work! this seems like a cleaner solution than the single-key exception we discussed before. just had minor comments.

lgtm, but since @stevendanna proposed this idea, let's see if he has any thoughts as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131750#-:-O8CpvlA6c8QKHi_tUQN:b3kh4h8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/kv/kvpb/api.proto` line 3538 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131750#-O8CoxR64dHf_YkL7BnZ:-O8CoxR64dHf_YkL7Bn_:b-f7pr9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/30b174ed445bc31e5f9f6c09f13a0aecfb53828d/pkg/kv/kvpb/api.proto#L3538)):*
> ```proto
>   // cluster will be subject to throttling. Since secondary tenants cannot access the KV
>   // store where this is saved, the value is propagated through the tenant settings.
>   // This is only set when the event type is METADATA_EVENT.
> ```

super nit: this comment could also mention that the value is seconds since epoch. i saw that you had the comment on the `GetClusterInitGracePeriodTS()` function, but it could help to have this in the comment for the data object as well.
___
*[`pkg/server/node.go` line 591 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131750#-O8CpUkG0uht5p99vjCx:-O8CpUkG0uht5p99vjCy:b7tcpmk) ([raw file](https://github.com/cockroachdb/cockroach/blob/30b174ed445bc31e5f9f6c09f13a0aecfb53828d/pkg/server/node.go#L591)):*
> ```Go
> 	spanConfigReporter spanconfig.Reporter,
> 	proxySender kv.Sender,
> 	licenseEnforcer *license.Enforcer,
> ```

just out of curiosity, do we need to pass the whole enforcer struct here, or would this be simpler if we passed the grace period `time.Time` only?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131750)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/kv/kvpb/api.proto` line 3538 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131750#-O8CoxR64dHf_YkL7BnZ:-O8D7QFW2-PHWG7IHmDG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30b174ed445bc31e5f9f6c09f13a0aecfb53828d/pkg/kv/kvpb/api.proto#L3538)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: this comment could also mention that the value is seconds since epoch. i saw that you had the comment on the `GetClusterInitGracePeriodTS()` function, but it could help to have this in the comment for the data object as well.
</blockquote></details>

Done.
___
*[`pkg/server/node.go` line 591 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131750#-O8CpUkG0uht5p99vjCx:-O8D6Ga40CqmAL-xYNob:bdnkws6) ([raw file](https://github.com/cockroachdb/cockroach/blob/30b174ed445bc31e5f9f6c09f13a0aecfb53828d/pkg/server/node.go#L591)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

just out of curiosity, do we need to pass the whole enforcer struct here, or would this be simpler if we passed the grace period `time.Time` only?
</blockquote></details>

Good point. However, the license enforcer isn’t fully initialized yet, so if we were to request the timestamp when `NewNode` is called, it wouldn’t reflect the value in the KV. We need to complete the system tenant setup first. After that, the secondary tenant will RPC back to the system tenant to request the timestamp, which will then be the correct value from the KV.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
Nice! Thanks for looking into this idea so quickly.  Overall the code looks good.  

My one question is how we get an updated value if we happen to get 0 when we first connect. It seems that there would be two cases where that could happen unless I've read the code wrong:

1) We are connecting to a KV node with an older binary version that doesn't send this value.
2) We asked for the grace period before it was cached (rare).

For (1), I'm assuming that right now that is perhaps only handled when the KV node restarts. I _think_ that would result in us getting the new value when the setting watcher retries.

For (2) it seems pretty rare, I wonder if perhaps just somehow waiting for this value to be non-zero when initially sending the tenant info might work.

I leave it to y'alls judgement on whether these are worth addressing now or in a follow up (assuming they are even issues).

spilchen reviewed (commented):
I believe only the first case would occur.

The second scenario shouldn’t happen because the system tenant’s startup should be complete by the time the secondary tenant requests metadata from the tenant connector. Unlike other startup tasks that usually run asynchronously, the license enforcer starts synchronously. This ensures that the system tenant has a cached copy of the timestamp from the KV.

As for the first case, I could add some retry logic. Maybe we can retry each time the grace period is requested when no license is present. To implement this, I’d need to cache a copy of the tenant connector.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131750#-:-O8H7231-U5dSXGTRh_o:b-ozkcjm)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
the latest change with retries looks good to me

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131750#-:-O8J1-7x-Vpi3ilq7t1m:bt15ymn)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131750)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/server/license/enforcer.go` line 320 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131750#-O8J2KO83oIo_ctkKhsa:-O8J2KO9-2m0Zie-knPc:b1fg3yb) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe40f89e7fb2f7881d1c14860384cbe0f5318587/pkg/server/license/enforcer.go#L320)):*
> ```Go
> // period as stored during cluster init. This is intended to be used a grace
> // period for when no license is installed.
> func (e *Enforcer) GetClusterInitGracePeriodEndTS() time.Time {
> ```

actually i just thought of a question: how frequently is this `GetClusterInitGracePeriodEndTS` function called?

we wouldn't want it to be too frequent, or too infrequent.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131750)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/server/license/enforcer.go` line 320 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131750#-O8J2KO83oIo_ctkKhsa:-O8J39PQ2NyJNLxQR1c2:b-71us1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe40f89e7fb2f7881d1c14860384cbe0f5318587/pkg/server/license/enforcer.go#L320)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

actually i just thought of a question: how frequently is this `GetClusterInitGracePeriodEndTS` function called?

we wouldn't want it to be too frequent, or too infrequent.
</blockquote></details>

Good question. This function is called during initialization and with every license change, which might not be frequent enough. Another logical point to check would be during throttle assessments. This could work if we only check when there's no license installed, but it might still occur too often. However, I believe the overhead of retrieving the timestamp from the tenant connector is minimal since it simply returns a cached copy and doesn't initiate any RPC calls.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131750)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/server/license/enforcer.go` line 320 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131750#-O8J2KO83oIo_ctkKhsa:-O8MYuUp5A5UXXYsv0w8:b-590vuq) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe40f89e7fb2f7881d1c14860384cbe0f5318587/pkg/server/license/enforcer.go#L320)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Good question. This function is called during initialization and with every license change, which might not be frequent enough. Another logical point to check would be during throttle assessments. This could work if we only check when there's no license installed, but it might still occur too often. However, I believe the overhead of retrieving the timestamp from the tenant connector is minimal since it simply returns a cached copy and doesn't initiate any RPC calls.
</blockquote></details>

I moved the polling into the threshold check, allowing us to retrieve any updated cluster initialization grace period timestamp sooner. As mentioned earlier, I’m not too concerned about the additional overhead since it’s just a cache lookup in the tenant connector. And this is only relevant when the timestamp wasn’t available during initialization.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (approved):
Thanks for going the extra mile here.

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131750)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/server/license/enforcer.go` line 320 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131750#-O8J2KO83oIo_ctkKhsa:-O8N3GBt4GboM_k7lxdz:bte4qu8) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe40f89e7fb2f7881d1c14860384cbe0f5318587/pkg/server/license/enforcer.go#L320)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I moved the polling into the threshold check, allowing us to retrieve any updated cluster initialization grace period timestamp sooner. As mentioned earlier, I’m not too concerned about the additional overhead since it’s just a cache lookup in the tenant connector. And this is only relevant when the timestamp wasn’t available during initialization.
</blockquote></details>

sounds good to me. does that mean access to `e.metadataAccessor` needs to be synchronized?


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r=rafiss,stevendanna

spilchen commented:
bors r-

spilchen committed:
server/license: Flow cluster init timestamp through tenant connector

Currently, secondary tenants cannot access the cluster initialization
grace period timestamp in the KV store because they lack permission to
access the system keyspace. Instead of reading it directly from the KV,
we now pass the timestamp through the tenant connector, making it
accessible to secondary tenants. The timestamp is only written once
during the system tenant’s initial startup, so it remains a constant
value from the perspective of secondary tenants. The value is sent from
the system tenant to secondary tenants using the existing
TenantSettingsRequest for METADATA_EVENT.

This change will be backported to versions 24.2, 24.1, 23.2, and 23.1.

Epic: CRDB-39988
Closes CRDB-42309
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131750)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/server/license/enforcer.go` line 320 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131750#-O8J2KO83oIo_ctkKhsa:-O8NnCRl7v4ICrwl2eNG:b-cwoze9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe40f89e7fb2f7881d1c14860384cbe0f5318587/pkg/server/license/enforcer.go#L320)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

sounds good to me. does that mean access to `e.metadataAccessor` needs to be synchronized?
</blockquote></details>

Good catch. I just added that to an `atomic.Pointer`.


<!-- Sent from Reviewable.io -->

spilchen commented:
bors r=rafiss,stevendanna

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131750)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/server/license/enforcer.go` line 103 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131750#-O8OdNNME0aPY3rhAEub:-O8OdNNME0aPY3rhAEuc:becul3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/163a755ae6a95ae7fbc21b011817f259724dd067/pkg/server/license/enforcer.go#L103)):*
> ```Go
> 	// in the metadata accessor. This is set to true during initialization if the
> 	// latest timestamp was not received.
> 	continueToPollMetadataAccessor atomic.Bool
> ```

with the number of atomic fields in this struct, i wonder if it would be easier to manage if all the fields were just put behind the `mutex`. we can discuss this more next week.


<!-- Sent from Reviewable.io -->

spilchen commented:
Note: I only backported this to 23.2, 24.1 and 24.2. This change depends on the metadata message in the tenant connector to send license information, which isn’t present in 23.1. Implementing this in 23.1 would require a different approach. Given that initial support for serverless was added in #131833 and there are currently no 23.1 serverless clusters in CC, I won’t be backporting this to 23.1.
